SQLite Exception during archive import

### Describe the bug


<!-- A clear and concise description of what the bug is. -->

During the import of an archive I created I got the following exception:

```
Report: starting import: pseudos.aiida
Report: Parameters
-------------------------------  ----------------
Archive                          pseudos.aiida
New Node Extras                  keep
Merge Node Extras (in database)  (k)eep
Merge Node Extras (in archive)   do (n)ot create
Merge Node Extras (in both)      (l)eave existing
Merge Comments                   leave
Computer Authinfos               exclude

Report: Skipping 1 existing User(s)
Report: Collecting Node(s) ...
Report: Skipping 1 existing Node(s)
Report: Adding 69 new node(s)
Report: Adding 1 new group(s)                                                                                                                                                                                                                                  
Report: Adding 70 Node(s) to new Group(s)                                                                                                                                                                                                                      
Report: Created new import Group: PK=5, label=20221101-151034                                                                                                                                                                                                  
Report: Checking keys against repository ...                                                                                                                                                                                                                   
Report: Skipping 1 existing repository files
Report: Adding 69 new repository files
63/6911/01/2022 03:10:44 PM <8935> sqlalchemy.pool.impl.NullPool: [ERROR] Exception during reset or similar
Traceback (most recent call last):
  File "/home/arnold/Documents/PhD/projects/aiida-venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 739, in _finalize_fairy
    fairy._reset(pool)
  File "/home/arnold/Documents/PhD/projects/aiida-venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 988, in _reset
    pool._dialect.do_rollback(self)
  File "/home/arnold/Documents/PhD/projects/aiida-venv/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 682, in do_rollback
    dbapi_connection.rollback()
sqlite3.ProgrammingError: SQLite objects created in a thread can only be used in that same thread. The object was created in thread id 140676517820224 and this is thread id 140676480427712.
11/01/2022 03:10:44 PM <8935> sqlalchemy.pool.impl.NullPool: [ERROR] Exception closing connection <sqlite3.Connection object at 0x7ff1aee9b3f0>
Traceback (most recent call last):
  File "/home/arnold/Documents/PhD/projects/aiida-venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 739, in _finalize_fairy
    fairy._reset(pool)
  File "/home/arnold/Documents/PhD/projects/aiida-venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 988, in _reset
    pool._dialect.do_rollback(self)
  File "/home/arnold/Documents/PhD/projects/aiida-venv/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 682, in do_rollback
    dbapi_connection.rollback()
sqlite3.ProgrammingError: SQLite objects created in a thread can only be used in that same thread. The object was created in thread id 140676517820224 and this is thread id 140676480427712.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/arnold/Documents/PhD/projects/aiida-venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 247, in _close_connection
    self._dialect.do_close(connection)
  File "/home/arnold/Documents/PhD/projects/aiida-venv/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 688, in do_close
    dbapi_connection.close()
sqlite3.ProgrammingError: SQLite objects created in a thread can only be used in that same thread. The object was created in thread id 140676517820224 and this is thread id 140676480427712.
Report: Committing transaction to database...                                                                                                                                                                                                                  
Success: imported archive pseudos.aiida
```
It does report a success at the end of the import, and all the nodes in the archive seem to be present after the import, but this exception makes me a bit worried that maybe something is corrupted in the database now.

Is there anyway to check whether the profile's database is not corrupted?

### Steps to reproduce

I don't know how to reproduce the error. This is the first time I get this exception during an import of an archive. I tried reproducing it by creating a clean new profile and importing the same archive there, but then the import completed successfully without any exceptions. I might try to delete the nodes that were imported and import them again in the same profile to see if it reproduces the error, but I haven't tried that yet.

### Expected behavior

<!-- A clear and concise description of what you expected to happen. -->
No sqlite exceptions during import

### Your environment

- Operating system [e.g. Linux]: Arch Linux (6.0.2-arch1-1)
- Python version [e.g. 3.7.1]: Python 3.9.13
- aiida-core version [e.g. 1.2.1]: 2.0.4
- PostgreSQL 14.5
- RabbitMQ 3.10.7
- aiida-pseudo  0.7.0
- aiida-siesta    2.0.0

--- a/src/aiida/storage/sqlite_dos/backend.py
+++ b/src/aiida/storage/sqlite_dos/backend.py
@@ -10,7 +10,7 @@
 
 from __future__ import annotations
 
-from functools import cached_property
+from functools import cached_property, lru_cache
 from pathlib import Path
 from shutil import rmtree
 from typing import TYPE_CHECKING, Optional
@@ -34,6 +34,7 @@ from ..psql_dos import PsqlDosBackend
 from ..psql_dos.migrator import REPOSITORY_UUID_KEY, PsqlDosMigrator
 
 if TYPE_CHECKING:
+    from aiida.orm.entities import EntityTypes
     from aiida.repository.backend import DiskObjectStoreRepositoryBackend
 
 __all__ = ('SqliteDosStorage',)
@@ -208,3 +209,19 @@ class SqliteDosStorage(PsqlDosBackend):
     @cached_property
     def users(self):
         return orm.SqliteUserCollection(self)
+
+    @staticmethod
+    @lru_cache(maxsize=18)
+    def _get_mapper_from_entity(entity_type: 'EntityTypes', with_pk: bool):
+        """Return the Sqlalchemy mapper and fields corresponding to the given entity.
+
+        :param with_pk: if True, the fields returned will include the primary key
+        """
+        from sqlalchemy import inspect
+
+        from ..sqlite_zip.models import MAP_ENTITY_TYPE_TO_MODEL
+
+        model = MAP_ENTITY_TYPE_TO_MODEL[entity_type]
+        mapper = inspect(model).mapper  # type: ignore[union-attr]
+        keys = {key for key, col in mapper.c.items() if with_pk or col not in mapper.primary_key}
+        return mapper, keys

--- a/src/aiida/storage/sqlite_zip/models.py
+++ b/src/aiida/storage/sqlite_zip/models.py
@@ -158,21 +158,23 @@ DbNode.user = sa_orm.relationship(
     ),
 )
 
+MAP_ENTITY_TYPE_TO_MODEL = {
+    EntityTypes.USER: DbUser,
+    EntityTypes.AUTHINFO: DbAuthInfo,
+    EntityTypes.GROUP: DbGroup,
+    EntityTypes.NODE: DbNode,
+    EntityTypes.COMMENT: DbComment,
+    EntityTypes.COMPUTER: DbComputer,
+    EntityTypes.LOG: DbLog,
+    EntityTypes.LINK: DbLink,
+    EntityTypes.GROUP_NODE: DbGroupNodes,
+}
+
 
 @functools.lru_cache(maxsize=10)
 def get_model_from_entity(entity_type: EntityTypes) -> Tuple[Any, Set[str]]:
     """Return the Sqlalchemy model and column names corresponding to the given entity."""
-    model = {
-        EntityTypes.USER: DbUser,
-        EntityTypes.AUTHINFO: DbAuthInfo,
-        EntityTypes.GROUP: DbGroup,
-        EntityTypes.NODE: DbNode,
-        EntityTypes.COMMENT: DbComment,
-        EntityTypes.COMPUTER: DbComputer,
-        EntityTypes.LOG: DbLog,
-        EntityTypes.LINK: DbLink,
-        EntityTypes.GROUP_NODE: DbGroupNodes,
-    }[entity_type]
+    model = MAP_ENTITY_TYPE_TO_MODEL[entity_type]
     mapper = sa.inspect(model).mapper
     column_names = {col.name for col in mapper.c.values()}
     return model, column_names

--- a/tests/storage/sqlite_dos/test_backend.py
+++ b/tests/storage/sqlite_dos/test_backend.py
@@ -12,3 +12,16 @@ def test_model():
     filepath = pathlib.Path.cwd() / 'archive.aiida'
     model = SqliteDosStorage.Model(filepath=filepath.name)
     assert pathlib.Path(model.filepath).is_absolute()
+
+
+def test_archive_import(aiida_config, aiida_profile_factory):
+    """Test that archives can be imported."""
+    from aiida.orm import Node, QueryBuilder
+    from aiida.tools.archive.imports import import_archive
+
+    from tests.utils.archives import get_archive_file
+
+    with aiida_profile_factory(aiida_config, storage_backend='core.sqlite_dos'):
+        assert QueryBuilder().append(Node).count() == 0
+        import_archive(get_archive_file('calcjob/arithmetic.add.aiida'))
+        assert QueryBuilder().append(Node).count() > 0

